src/v/cluster/rm_stm.cc

```diff
@@ -1199,91 +1193,27 @@ ss::future<result<kafka_result>> rm_stm::transactional_replicate(
-                if (barrier) {
-                    co_return errc::sequence_out_of_order;
-                }
-                co_return errc::not_leader;
-            }
-        } else if (pid_seq->second.entry.seq == bid.last_seq) {
-            if (pid_seq->second.entry.last_offset() == -1) {
-                if (pid_seq->second.term == synced_term) {
-                    vlog(
-                      _ctx_log.debug,
-                      "Status of the original attempt pid:{} seq:{} is "
-                      "unknown. Returning not_leader_for_partition to trigger "
-                      "retry.",
-                      bid.pid,
-                      bid.last_seq);
-                    co_return errc::not_leader;
-                } else {
-                    // when the term a tx originated on doesn't match the
-                    // current term it means that the re-election happens; upon
-                    // re-election we sync and it catches up with all records
-                    // replicated in previous term. since the cached offset for
-                    // a given seq is still -1 it means that the replication
-                    // hasn't passed so we updating the term to pretent that the
-                    // retry (seqs match) put it there
-                    pid_seq->second.term = synced_term;
-                }
-            } else {
-                vlog(
-                  _ctx_log.trace,
-                  "cache hit for pid:{} seq:{}",
-                  bid.pid,
-                  bid.last_seq);
-                co_return kafka_result{
-                  .last_offset = pid_seq->second.entry.last_offset};
-            }
-        } else {
-            for (auto& entry : pid_seq->second.entry.seq_cache) {
-                if (entry.seq == bid.last_seq) {
-                    if (entry.offset() == -1) {
-                        vlog(
-                          _ctx_log.error,
-                          "cache hit for pid:{} seq:{} resolves to -1 offset",
-                          bid.pid,
-                          entry.seq);
-                    }
-                    vlog(
-                      _ctx_log.trace,
-                      "cache hit for pid:{} seq:{}",
-                      bid.pid,
-                      entry.seq);
-                    co_return kafka_result{.last_offset = entry.offset};
-                }
-            }
-            if (!check_seq(bid, synced_term)) {
-                auto barrier = co_await _raft->linearizable_barrier();
-                if (barrier) {
-                    co_return errc::sequence_out_of_order;
-                }
-                co_return errc::not_leader;
-            }
-        }
-    } else {
-        // this is the first attempt in the tx, reset dedupe cache
-        reset_seq(bid, synced_term);
-
-        _mem_state.estimated[bid.pid] = last_applied_offset();
+    auto request = producer->enqueue_request(bid, synced_term);
```

Comment:
idempotency sequence number check differs from tx sequence number but you handle then as it's the same

Reply:
can you clarify a bit more where it differs?

Reply:
https://github.com/redpanda-data/redpanda/commit/8cd0bc66ed5214c706305c86ca4c11a8ffd192bd

Reply:
fixed.